Avoid graph break during NJT construction

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130505

Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. Today, this association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

At a high level, we want to be able to do things like this successfully:
```python
@torch.compile(fullgraph=True)
def f(values, offsets):
    nt = torch.nested.nested_tensor_from_jagged(values, offsets)
    return nt + 1

# called like: g(nt, torch.randn_like(nt.values()), nt.offsets())
@torch.compile(fullgraph=True)
def g(nt1, values2, offsets):
    nt2 = torch.nested.nested_tensor_from_jagged(values2, offsets)
    return nt1 + nt2
```

`fullgraph=True` is necessary for Ads inference on AOTI, for example.

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**Side note:** NJT's usage of a real ATen view op for construction bypasses other construction-based graph breaks that I believe other subclasses have run into.

**Long-term solution:** Use a union find to maintain equivalence relationships! This is a big overhaul and we need to unblock various use cases (Ads, preproc, etc.) ASAP.

**Short-term solution in this PR:** *Comprehensively* include `fake tensor -> symbolic nested int` entries within the nested int registry.
* Symbolicize any associated nested ints during `offsets` / `lengths` fake-ification for storage within the registry.
* For fake tensors that take part in NJT construction in-graph (but previously had no associated nested int), create a new (unbacked) symbolic nested int on-the-fly and associate it with the fake tensor in the registry.
* Handle conceptual equivalences between fake tensors within AOTAutograd (details below).
* Functional tensors contain fake tensors; unwrap them to store the associated `fake tensor -> symbolic nested int` in the registry.
* NB: #130292 moves away from nested int registry usage (in line with the longer-term solution), preferring to store symbolic nested ints directly onto fake tensors.

**PR Details:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.
* Introduces the concept of nested int registry *branching* to avoid poisoning the eager registry state with compile-specific entries post-compile.
    * Idea: branch the registry at the beginning of torch.compile, operate within the branched registry, revert to original state post-compile.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

jbschlosser committed:
Avoid graph break during NJT construction

[ghstack-poisoned]

jbschlosser commented on test/dynamo/test_subclasses.py:
```diff
@@ -1585,6 +1585,98 @@ def binary(nt1, nt2):
+        self.assertEqual(out, ref_out)
+        self.assertEqual(grad, ref_grad)
+
+        #
+        # Mixed case
+        #
+        nt, _ = self._get_jagged_tensor(((2, 3, 4), 5), None)
+        values = nt.values().detach() # requires_grad_(False)
+        torch._dynamo.mark_dynamic(values, 0)
+
+        def fn(nt, values, offsets):
+            nt2 = torch.nested.nested_tensor_from_jagged(values, offsets)
+            # return nt2 * 3
+            return torch.ops.aten.mul.Tensor(nt, nt2)
+            # return nt * nt2
+
+        out = torch.compile(fn, fullgraph=True, backend="aot_eager")(nt, values, nt.offsets())
```
Note: this case works now!

jbschlosser commented on torch/_dynamo/source.py:
```diff
@@ -285,6 +285,25 @@ def name(self):
+        assert self.base is not None
+
+    def reconstruct(self, codegen):
+        # This is wrong, when is this used?
+        return [
+            *self.base.reconstruct(codegen),
+            codegen.create_load_attr(self.prop),
+        ]
```
TODO: fix

jbschlosser commented on torch/_functorch/_aot_autograd/collect_metadata_analysis.py:
```diff
@@ -160,7 +160,10 @@ def inner(*flat_args):
         with disable_above, mode:
             # precondition: The passed in function already handles unflattening inputs + flattening outputs
             flat_f_args = pytree.tree_map(_to_fun, flat_args)
+
```
TODO: remove nested-specific logic from here

jbschlosser commented on torch/_subclasses/meta_utils.py:
```diff
@@ -818,6 +824,7 @@ def _empty_create_subclass(
                             )
                             assert t.data is not None
                             _safe_copy(r.real_tensor, t.data)
+                    r.nested_int = t.nested_int
```
it's bad that this propagation logic is in two places. Ideally I think `_empty_create_subclass()` would operate more recursively..

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -16,8 +16,38 @@ def get_tensor_symint(tensor, *, coeff=1):
+        if isinstance(tensor, FakeTensor):
+            shape_env = tensor.fake_mode.shape_env
+
+            if tensor.source is None:
+                # Source is None in two cases:
+                # (1) tensor._base is _dummy_instance OR
+                # (2) tensor is an intermediate
+                src = torch._dynamo.source.EphemeralSource("intermediate_offsets")
+            else:
+                src = torch._dynamo.source.NestedIntSource(tensor.source)
+
+            tensor_symint = shape_env.create_symintnode(
+                sym=shape_env.create_symbol(
+                    val=tensor_symint,
+                    source=src,
+                ),
+                hint=tensor_symint,
+                source=src,
+            )
```
TODO: factor this out from here and fake-ification

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

soulitzer commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -411,6 +411,32 @@ def is_concrete_bool(a: Union[bool, SymBool]) -> bool:
+        else:
+            nested_int = nested_int.node.hint
+
+    nested_source: Source = EphemeralSource("intermediate_offsets_or_lengths")
```
Nice! Always using EphemeralSource sounds like a good idea for now to avoid having to thinking about saving source on Tensor until we actually need to.

jbschlosser replied:
exactly! booted that til later for the best chance of getting something landable now. none of the test cases we currently have try to guard on a nested int, so we don't need a non-ephemeral source

soulitzer commented on torch/nested/_internal/nested_tensor.py:
```diff
@@ -1,23 +1,49 @@
+    try:
+        yield
+    finally:
+        _tensor_id_counter = orig_counter
```
I made some updates to the other PR btw, and one of the changes was adding a new separate tensor_id_counter specific for tracing  to avoid polluting any eager state while tracing (more info in the newly beefed up PR descr)

For now I'm storing it on FakeTensorMode and resetting based on the value of the eager id counter upon enter/exit, but there are probably more subtleties I'm missing.

jbschlosser replied:
ooh I like this much better! lemme go check it out

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser committed:
Update on "Avoid graph break during NJT construction"


Based on #130292; playing around with different designs, etc.

**Background:** Nested ints uniquely define ragged structure and are associated with metadata tensors `offsets` / `lengths`. This association is maintained within the nested int registry, which maps `offsets` / `lengths` -> `nested int`. If a new `offsets` / `lengths` tensor object is seen during NJT construction, a new nested int (e.g. `j0`) is created to associate with it, and the nested int counter is incremented (next up: `j1`).

**Core problem:** During tracing, various fake tensors and functional tensors are created that *conceptually* refer to the same object as their real source. For fake / functional `offsets` / `lengths`, these should be considered to have the same ragged structure as the real `offsets` / `lengths`, and thus be associated with the same nested int. Further, whenever we're dealing with fake tensors, we should be dealing with symbolic nested ints. These are not done consistently today, leading to all sorts of problems (#126472, #130272, [error](https://gist.github.com/davidberard98/877a52f6ea57025cc122d64361e598da)). To avoid graph breaks or hard errors during in-graph NJT construction, this needs to be cleaned up.

**This PR:**
* During dense tensor fake-ification, mirrors `offsets -> nested int` relationships via `fake offsets -> symbolic nested int` entries within the nested int registry. Details:
    * `describe_tensor()` is updated to store any nested int associated with the real dense tensor as part of its `MetaTensorDesc`.
    * During dense tensor fake-ification, if the tensor's desc has a nested int, it is symbolicized within the relevant ShapeEnv.
    * An `output fake tensor -> symbolic nested int` entry is added to the nested int registry.
        * Note: this is done in two places right now: the usual end of dense tensor fake-ification and within `empty_create_subclass()` (which I think should call `meta_tensor()` recursively instead of what it does now).
* Updates AOTAutograd in a couple places to achieve the same conceptual grouping of `offsets` / `lengths` with associated fake intermediates:
    * During metadata collection (which is run twice), this PR emulates idempotence by restoring the nested int counter state after the fw graph run. Any `offsets` / `lengths` -> `nested int` relationships (and their fake analogues) should have the same nested int values across fw graph runs. This is accomplished via a (somewhat hacky) `freeze_nested_int_counter()` context manager.
    * As per `Note [AOT Autograd: Views to avoid tangents aliasing inputs]`, `view_avoid_dupes_with_primals()` purposefully creates an aliased output via `out = t.view(t.shape)`. Conceptually, `out` here should be associated with the same ragged structure as `t`, so this PR updates the nested int registry to maintain this relationship.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

jbschlosser commented:
As discussed offline, closing in favor of #130292.
